Notion API + Google Calendar Events Feature Microservice

ashworks1706 opened the pull request:
This PR introduces a robust microservice that enables seamless synchronization between Notion databases and Google Calendar events.

## Key Features

- Real-time event synchronization from Notion to Google Calendar
- Webhook support for automatic updates
- Comprehensive event tracking and change detection
- Support for event locations, descriptions, and attendees
- Timezone handling (America/Phoenix)
- Detailed logging system

## Technical Implementation

- Flask-based webhook endpoint
- Google Calendar API integration with service account authentication
- Notion API integration with database querying
- JSON-based state management for event tracking
- Error handling with detailed logging
- Configurable through appConfig.json

## Changes Made

- Added core synchronization logic
- Implemented webhook handlers
- Added event parsing and mapping
- Integrated logging system
- Added configuration management
- Implemented error handling

## Testing Instructions

1. Configure appConfig.json with your credentials
2. Start the Flask server
3. Configure Notion webhook to point to /notion-webhook
4. Create/modify events in Notion to test synchronization

## Security Considerations

- Credentials stored in separate configuration file
- Error logging with sensitive data protection
- API scope restrictions for minimal access
- Webhook authentication ready


This PR addresses the need for automated calendar synchronization while maintaining security and scalability standards.

ashworks1706 commented:
![image](https://github.com/user-attachments/assets/eca21b10-f1a8-49aa-ad1e-cbc5e0555cd5)

![image](https://github.com/user-attachments/assets/8187e0a7-1bd3-4905-9f09-d2865d8dd42d)

![image](https://github.com/user-attachments/assets/47fb0d11-81c1-42cf-b4d8-48df8ad2caeb)

![image](https://github.com/user-attachments/assets/15b96b1d-6f92-4555-9b40-ce5d9f6fabc5)

![image](https://github.com/user-attachments/assets/cc6119e9-4644-4da5-a68a-3262e0d96a47)

![image](https://github.com/user-attachments/assets/03c41cdf-8a45-4973-89ec-c31f0a22debe)

![image](https://github.com/user-attachments/assets/0e0b9a6b-8639-4b9a-857e-0fc3f6a99df6)

### Mentioning a guest won't work since this is not an organizational account and Domain wide permission is required to add guests to events, which can be only done with a google organizational account as super admin

#### Always seperate guest's emails with comma
![image](https://github.com/user-attachments/assets/b7915db8-4137-4fce-8fd3-277985238788)

![image](https://github.com/user-attachments/assets/bd200bf3-dba2-4757-9e38-438a0aa723ef)

benjuntilla commented:
instead of creating a completely new subdirectory `soda-backend-api`, this code and documentation should be put in `modules` along with all the other existing endpoints. then, register this as a blueprint in `main.py`. this will enforce modularity and consistency.

i still need to do an in-depth code review and functional requirements test.

ashworks1706 commented:
> instead of creating a completely new subdirectory `soda-backend-api`, this code and documentation should be put in `modules` along with all the other existing endpoints. then, register this as a blueprint in `main.py`. this will enforce modularity and consistency.
> 
> 
> 
> i still need to do an in-depth code review and functional requirements test. 

gotcha i figured you guys can do that since i did not have good understanding of the codebase, ill do it rightaway

ashworks1706 commented:
**Update Summary (February 1, 2025):**

The `calendar` module has been successfully updated and integrated into the SoDA internal API following the outlined requirements. Key changes include:

1. **Environment Variable Configuration:**
   - Replaced `appConfig.json` with `.env` for secure and scalable configuration management.
   - Added a `.env.template` file to guide environment variable setup.

2. **Blueprint Integration:**
   - The `calendar` module is now registered as a blueprint in `main.py`, ensuring modularity and consistency with the existing project structure.

3. **Enhanced Functionality:**
   - Added robust Google Calendar and Notion API integrations with comprehensive error handling.
   - Implemented event parsing, synchronization, and attendee management.
   - Introduced functionality to clear outdated or future events automatically.

4. **Improved Documentation:**
   - Updated the README to reflect the new `.env`-based configuration and usage instructions.
   - Added detailed steps for setting up Notion databases, Google Calendar permissions, and running the service.

5. **Security Enhancements:**
   - Credentials are now securely managed via environment variables.
   - Sensitive data is protected from accidental exposure in version control.


### Next Steps:
- Please review the updated `calendar` module and ensure it meets all functional requirements.
- Test the webhook endpoint (`/calendar/notion-webhook`) for both GET (health check) and POST (event synchronization) requests.
- Verify that environment variables are correctly set up using `.env.template`.

code-wolf-byte commented:
1. Function def `validate_environment()` and variable `CONFIG` are redundant as they should be handled by  `modules\utils\config.py`
2. The current implementation creates a new server which will mess with the production build of the application, the section
```py
if __name__ == '__main__':
    app.run(
        debug=CONFIG['server']['debug'],
        port=CONFIG['server']['port']
    )

```
should be removed. 

3. Google API keys would be stored at the project's root and should be imported relatively. 
4. Type hinting is random and should either be completely added or completely removed. 
5. The functions are relatively undocumented and should be documented for future extendibility.

ashworks1706 commented:
Fixed Issues-

1. Created notion and logging instance on shared.py instead and imported.
![{AC10389F-2D37-437D-B1C8-F582BCA4C59E}](https://github.com/user-attachments/assets/e5161d00-c31e-4f99-ab90-abf9372b37d1)
![{5A32F967-EDD6-4526-A37A-77578BE48EA7}](https://github.com/user-attachments/assets/ef2e8cd4-4457-4ee8-a952-523a5a7e0fe3)


2. Removed `validate_environment()` function and instead added env variables in config.py
![{832614C7-14E9-41DC-9515-B0E439F61BDD}](https://github.com/user-attachments/assets/de962d07-b56c-4cb0-9059-88111a81d72f)

3. Added more function definitions and type hinting
![{E4365980-5BB6-4EA4-8952-94CCBC775D3A}](https://github.com/user-attachments/assets/df3df9cc-7d89-4b0b-a68c-178016775209)
